File: VoiceType/Services/AsrService.swift
import Foundation

enum AsrError: LocalizedError {
    case modelMissing
    case vadMissing

    var errorDescription: String? {
        switch self {
        case .modelMissing:
            return "本地识别模型未安装（\(SettingsStore.localAsrModel.label)），请在项目目录运行 "
                + SettingsStore.localAsrModel.installCommand
        case .vadMissing:
            return "VAD 模型未安装，请运行 ./scripts/export_model.sh 或 ./scripts/download_models.sh"
        }
    }
}

enum ModelPaths {
    static var modelsDir: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("VoiceType/models", isDirectory: true)
    }

    // MARK: - SenseVoice（历史布局，位于 models/ 根）

    static var asrModel: URL { modelsDir.appendingPathComponent("model.int8.onnx") }
    static var tokens: URL { modelsDir.appendingPathComponent("tokens.txt") }

    // MARK: - Fun-ASR-Nano-2512（models/funasr-nano/）

    static var funasrDir: URL {
        modelsDir.appendingPathComponent("funasr-nano", isDirectory: true)
    }
    static var funasrEncoder: URL {
        funasrDir.appendingPathComponent("encoder_adaptor.int8.onnx")
    }
    static var funasrLLM: URL { funasrDir.appendingPathComponent("llm.int8.onnx") }
    static var funasrEmbedding: URL { funasrDir.appendingPathComponent("embedding.int8.onnx") }
    /// tokenizer 目录（vocab.json / merges.txt / tokenizer.json）
    static var funasrTokenizer: URL {
        funasrDir.appendingPathComponent("Qwen3-0.6B", isDirectory: true)
    }

    // MARK: - Qwen3-ASR-0.6B（models/qwen3-asr/）

    static var qwen3Dir: URL {
        modelsDir.appendingPathComponent("qwen3-asr", isDirectory: true)
    }
    static var qwen3ConvFrontend: URL { qwen3Dir.appendingPathComponent("conv_frontend.onnx") }
    static var qwen3Encoder: URL { qwen3Dir.appendingPathComponent("encoder.int8.onnx") }
    static var qwen3Decoder: URL { qwen3Dir.appendingPathComponent("decoder.int8.onnx") }
    /// tokenizer 目录（vocab.json / merges.txt / tokenizer_config.json，无 tokenizer.json）
    static var qwen3Tokenizer: URL {
        qwen3Dir.appendingPathComponent("tokenizer", isDirectory: true)
    }

    // MARK: - 公共组件

    static var vadModel: URL { modelsDir.appendingPathComponent("silero_vad.onnx") }
    static var segmentationModel: URL { modelsDir.appendingPathComponent("segmentation.onnx") }
    static var speakerEmbeddingModel: URL {
        modelsDir.appendingPathComponent("speaker-embedding.onnx")
    }

    static var vadPresent: Bool {
        FileManager.default.fileExists(atPath: vadModel.path)
    }

    static var diarizationPresent: Bool {
        [segmentationModel, speakerEmbeddingModel].allSatisfy {
            FileManager.default.fileExists(atPath: $0.path)
        }
    }

    /// 指定档位的模型文件是否全部就位
    static func isPresent(_ model: LocalAsrModel) -> Bool {
        func exists(_ url: URL) -> Bool {
            FileManager.default.fileExists(atPath: url.path)
        }
        switch model {
        case .senseVoice:
            return exists(asrModel) && exists(tokens)
        case .funasrNano:
            return [funasrEncoder, funasrLLM, funasrEmbedding].allSatisfy(exists)
                && ["vocab.json", "merges.txt", "tokenizer.json"].allSatisfy {
                    exists(funasrTokenizer.appendingPathComponent($0))
                }
        case .qwen3Asr:
            return [qwen3ConvFrontend, qwen3Encoder, qwen3Decoder].allSatisfy(exists)
                && ["vocab.json", "merges.txt", "tokenizer_config.json"].allSatisfy {
                    exists(qwen3Tokenizer.appendingPathComponent($0))
                }
        }
    }

    /// 当前选择的本地模型是否就绪
    static var allPresent: Bool { isPresent(SettingsStore.localAsrModel) }
}

/// 离线识别服务（sherpa-onnx）：SenseVoiceSmall / Fun-ASR-Nano-2512 / Qwen3-ASR-0.6B。
/// 推理在专用串行队列执行，模型常驻内存；切换档位后调用 reload() 重建。
final class AsrService: @unchecked Sendable {
    private var recognizer: SherpaOnnxOfflineRecognizer?
    private var loadedModel: LocalAsrModel?
    private let queue = DispatchQueue(label: "com.catsonkeyboard.voicetype.asr", qos: .userInitiated)

    /// LLM 型模型存在 max_total_len（≈512 token ≈ 20s 音频）截断上限，
    /// 超过该时长的单次转写先经 VAD 分段再逐段识别。
    private static let llmMaxDirectSeconds: Double = 20

    private static func isLLM(_ model: LocalAsrModel) -> Bool {
        switch model {
        case .funasrNano, .qwen3Asr: return true
        case .senseVoice: return false
        }
    }

    /// 预热：后台加载当前档位模型（App 启动时调用）
    func warmUp() {
        queue.async { _ = try? self.loadedRecognizer() }
    }

    /// 切换档位后调用：释放旧模型并立即按新档位重载
    func reload() {
        queue.async {
            self.recognizer = nil
            self.loadedModel = nil
            _ = try? self.loadedRecognizer()
        }
    }

    private func loadedRecognizer() throws -> SherpaOnnxOfflineRecognizer {
        let model = SettingsStore.localAsrModel
        if let recognizer, loadedModel == model { return recognizer }
        guard ModelPaths.isPresent(model) else { throw AsrError.modelMissing }


        let modelConfig: SherpaOnnxOfflineModelConfig
        switch model {
        case .senseVoice:
            modelConfig = sherpaOnnxOfflineModelConfig(
                tokens: ModelPaths.tokens.path,
                numThreads: 4,
                senseVoice: sherpaOnnxOfflineSenseVoiceModelConfig(
                    model: ModelPaths.asrModel.path,
                    language: "auto",
                    useInverseTextNormalization: true
                ))
        case .funasrNano:
            modelConfig = sherpaOnnxOfflineModelConfig(
                tokens: "",
                numThreads: 4,
                funasrNano: sherpaOnnxOfflineFunASRNanoModelConfig(
                    encoderAdaptor: ModelPaths.funasrEncoder.path,
                    llm: ModelPaths.funasrLLM.path,
                    embedding: ModelPaths.funasrEmbedding.path,
                    tokenizer: ModelPaths.funasrTokenizer.path,
                    maxNewTokens: 512
                ))
        case .qwen3Asr:
            modelConfig = sherpaOnnxOfflineModelConfig(
                tokens: "",
                numThreads: 4,
                qwen3Asr: sherpaOnnxOfflineQwen3ASRModelConfig(
                    convFrontend: ModelPaths.qwen3ConvFrontend.path,
                    encoder: ModelPaths.qwen3Encoder.path,
                    decoder: ModelPaths.qwen3Decoder.path,
                    tokenizer: ModelPaths.qwen3Tokenizer.path,
                    maxNewTokens: 512
                ))
        }
        var config = sherpaOnnxOfflineRecognizerConfig(
            featConfig: sherpaOnnxFeatureConfig(),
            modelConfig: modelConfig
        )
        let r = SherpaOnnxOfflineRecognizer(config: &config)
        recognizer = r
        loadedModel = model
        return r
    }

    /// 单段音频转写（16kHz 单声道）。LLM 模型超 20s 时自动 VAD 分段。
    func transcribe(samples: [Float], sampleRate: Int = 16000) async throws -> String {
        try await withCheckedThrowingContinuation { cont in
            queue.async {
                do {
                    let text = try self.transcribeSync(samples: samples, sampleRate: sampleRate)
                    cont.resume(returning: text)
                } catch {
                    cont.resume(throwing: error)
                }
            }
        }
    }

    private func transcribeSync(samples: [Float], sampleRate: Int) throws -> String {
        let r = try loadedRecognizer()
        let duration = Double(samples.count) / Double(max(sampleRate, 1))
        if Self.isLLM(SettingsStore.localAsrModel), duration > Self.llmMaxDirectSeconds {
            let pieces = try Self.vadTranscribe(
                samples: samples, recognizer: r, maxSpeechDuration: 15)
            return pieces.joined(separator: "\n")
        }
        return r.decode(samples: samples, sampleRate: sampleRate).text
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// 文件转写：解码 → silero VAD 分段 → 逐段识别，段间换行
    func transcribeFile(
        url: URL, onProgress: @escaping @Sendable (Double) -> Void
    ) async throws -> String {
        let samples = try AudioFileDecoder.decode16kMono(url: url)
        let maxSpeech: Float = Self.isLLM(SettingsStore.localAsrModel) ? 15 : 20
        return try await withCheckedThrowingContinuation { cont in
            queue.async {
                do {
                    let r = try self.loadedRecognizer()
                    let pieces = try Self.vadTranscribe(
                        samples: samples, recognizer: r, maxSpeechDuration: maxSpeech,
                        onProgress: onProgress)
                    cont.resume(returning: pieces.joined(separator: "\n"))
                } catch {
                    cont.resume(throwing: error)
                }
            }
        }
    }

    /// VAD 切分 + 逐段识别。LLM 模型 maxSpeechDuration 取 15s（防 max_total_len 截断），
    /// SenseVoice 沿用 20s。空段自动丢弃。
    private static func vadTranscribe(
        samples: [Float],
        recognizer: SherpaOnnxOfflineRecognizer,
        maxSpeechDuration: Float,
        onProgress: (@Sendable (Double) -> Void)? = nil
    ) throws -> [String] {
        guard ModelPaths.vadPresent else { throw AsrError.vadMissing }
        let silero = sherpaOnnxSileroVadModelConfig(
            model: ModelPaths.vadModel.path,
            threshold: 0.5,
            minSilenceDuration: 0.5,
            minSpeechDuration: 0.25,
            windowSize: 512,
            maxSpeechDuration: maxSpeechDuration
        )
        var vadConfig = sherpaOnnxVadModelConfig(sileroVad: silero)
        let vad = SherpaOnnxVoiceActivityDetectorWrapper(
            config: &vadConfig, buffer_size_in_seconds: 120)

        var pieces: [String] = []
        func drainSegments() {
            while !vad.isEmpty() {
                let seg = vad.front()
                let text = recognizer.decode(samples: seg.samples).text
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                if !text.isEmpty { pieces.append(text) }
                vad.pop()
            }
        }

        let window = 512
        var i = 0
        while i < samples.count {
            let end = min(i + window, samples.count)
            vad.acceptWaveform(samples: Array(samples[i..<end]))
            drainSegments()
            i = end
            onProgress?(Double(i) / Double(max(samples.count, 1)))
        }
        vad.flush()
        drainSegments()
        return pieces
    }
}

File: windows/src/VoiceType/Services/AsrService.cs
using VoiceType.Interop.SherpaOnnx;

namespace VoiceType.Services;

public sealed class AsrException : Exception
{
    public AsrException(string message) : base(message) { }
}

/// <summary>
/// 离线识别服务（sherpa-onnx）：SenseVoiceSmall / Fun-ASR-Nano-2512 / Qwen3-ASR-0.6B。
/// 推理在串行信号量保护下执行（对应 macOS 专用串行队列），模型常驻内存；
/// 切换档位后调用 Reload() 重建。
/// </summary>
public sealed class AsrService : IDisposable
{
    /// <summary>LLM 型模型存在 max_total_len（≈512 token ≈ 20s 音频）截断上限，
    /// 超过该时长的单次转写先经 VAD 分段再逐段识别。</summary>
    private const double LlmMaxDirectSeconds = 20;

    private readonly SemaphoreSlim _gate = new(1, 1);
    private readonly Func<SettingsStore> _settings;
    private OfflineRecognizer? _recognizer;
    private LocalAsrModel? _loadedModel;

    public AsrService(SettingsStore? settings = null)
    {
        _settings = () => settings ?? SettingsStore.Default;
    }

    private SettingsStore Settings => _settings();

    private static bool IsLlm(LocalAsrModel model) =>
        model is LocalAsrModel.FunasrNano or LocalAsrModel.Qwen3Asr;

    /// <summary>预热：后台加载当前档位模型（App 启动时调用）</summary>
    public void WarmUp() => _ = Task.Run(async () =>
    {
        await _gate.WaitAsync();
        try { _ = LoadedRecognizer(); }
        catch { /* 启动期模型未安装属正常态，不抛 */ }
        finally { _gate.Release(); }
    });

    /// <summary>切换档位后调用：释放旧模型并立即按新档位重载</summary>
    public async Task ReloadAsync()
    {
        await _gate.WaitAsync();
        try
        {
            _recognizer?.Dispose();
            _recognizer = null;
            _loadedModel = null;
            _ = LoadedRecognizer();
        }
        finally
        {
            _gate.Release();
        }
    }

    private OfflineRecognizer LoadedRecognizer()
    {
        LocalAsrModel model = Settings.LocalAsrModel;
        if (_recognizer is not null && _loadedModel == model)
            return _recognizer;
        if (!ModelPaths.IsPresent(model))
            throw new AsrException(
                $"本地识别模型未安装（{model.Label()}），请在项目目录运行 {model.InstallCommand()}");

        var options = new OfflineRecognizerOptions
        {
            ModelKind = model switch
            {
                LocalAsrModel.SenseVoice => OfflineModelKind.SenseVoice,
                LocalAsrModel.FunasrNano => OfflineModelKind.FunAsrNano,
                LocalAsrModel.Qwen3Asr => OfflineModelKind.Qwen3Asr,
                _ => throw new ArgumentOutOfRangeException(),
            },
            NumThreads = 4,
            Tokens = ModelPaths.Tokens,
            SenseVoice = new SenseVoiceOptions
            {
                Model = ModelPaths.AsrModel,
                Language = "auto",
                UseInverseTextNormalization = true,
            },
            FunAsrNano = new FunAsrNanoOptions
            {
                EncoderAdaptor = ModelPaths.FunasrEncoder,
                Llm = ModelPaths.FunasrLlm,
                Embedding = ModelPaths.FunasrEmbedding,
                TokenizerDir = ModelPaths.FunasrTokenizer,
                MaxNewTokens = 512,
            },
            Qwen3Asr = new Qwen3AsrOptions
            {
                ConvFrontend = ModelPaths.Qwen3ConvFrontend,
                Encoder = ModelPaths.Qwen3Encoder,
                Decoder = ModelPaths.Qwen3Decoder,
                TokenizerDir = ModelPaths.Qwen3Tokenizer,
                MaxNewTokens = 512,
            },
        };
        var r = OfflineRecognizer.Create(options);
        _recognizer = r;
        _loadedModel = model;
        return r;
    }

    /// <summary>单段音频转写（16kHz 单声道）。LLM 模型超 20s 时自动 VAD 分段。
    /// 推理在信号量串行 + 线程池执行，调用方（UI）不阻塞。</summary>
    public async Task<string> TranscribeAsync(
        float[] samples, int sampleRate = 16000, CancellationToken ct = default)
    {
        await _gate.WaitAsync(ct);
        try
        {
            return await Task.Run(() => TranscribeSync(samples, sampleRate), ct);
        }
        finally
        {
            _gate.Release();
        }
    }

    private string TranscribeSync(float[] samples, int sampleRate)
    {
        OfflineRecognizer r = LoadedRecognizer();
        double duration = samples.Length / (double)Math.Max(sampleRate, 1);
        if (IsLlm(Settings.LocalAsrModel) && duration > LlmMaxDirectSeconds)
        {
            List<string> pieces = VadTranscribe(samples, r, maxSpeechDuration: 15f, null);
            return string.Join("\n", pieces);
        }
        return r.Decode(samples, sampleRate).Trim();
    }

    /// <summary>文件转写：解码 → silero VAD 分段 → 逐段识别，段间换行</summary>
    public async Task<string> TranscribeFileAsync(
        string path, Action<double>? onProgress = null, CancellationToken ct = default)
    {
        float[] samples = AudioFileDecoder.Decode16kMono(path);
        float maxSpeech = IsLlm(Settings.LocalAsrModel) ? 15f : 20f;
        await _gate.WaitAsync(ct);
        try
        {
            OfflineRecognizer r = LoadedRecognizer();
            List<string> pieces = await Task.Run(
                () => VadTranscribe(samples, r, maxSpeech, onProgress), ct);
            return string.Join("\n", pieces);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>VAD 切分 + 逐段识别。LLM 模型 maxSpeechDuration 取 15s（防 max_total_len 截断），
    /// SenseVoice 沿用 20s。空段自动丢弃。</summary>
    private static List<string> VadTranscribe(
        float[] samples, OfflineRecognizer recognizer, float maxSpeechDuration,
        Action<double>? onProgress)
    {
        if (!ModelPaths.VadPresent)
            throw new AsrException("VAD 模型未安装，请运行 .\\windows\\scripts\\download_models.ps1 或 export_sensevoice.ps1");
        using var vad = VoiceActivityDetector.Create(new VadOptions
        {
            Model = ModelPaths.VadModel,
            Threshold = 0.5f,
            MinSilenceDuration = 0.5f,
            MinSpeechDuration = 0.25f,
            WindowSize = 512,
            MaxSpeechDuration = maxSpeechDuration,
            BufferSizeInSeconds = 120,
        });

        var pieces = new List<string>();
        void DrainSegments()
        {
            while (!vad.IsEmpty())
            {
                SpeechSegmentData seg = vad.Front();
                string text = recognizer.Decode(seg.Samples).Trim();
                if (text.Length > 0)
                    pieces.Add(text);
                vad.Pop();
            }
        }

        const int window = 512;
        int i = 0;
        while (i < samples.Length)
        {
            int end = Math.Min(i + window, samples.Length);
            vad.AcceptWaveform(samples[i..end]);
            DrainSegments();
            i = end;
            onProgress?.Invoke(i / (double)Math.Max(samples.Length, 1));
        }
        vad.Flush();
        DrainSegments();
        return pieces;
    }

    public void Dispose()
    {
        _recognizer?.Dispose();
        _recognizer = null;
        _loadedModel = null;
        _gate.Dispose();
    }
}
